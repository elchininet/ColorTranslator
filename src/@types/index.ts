export type NumberOrString = number | string;

export interface HEXObject {
    r: string;
    g: string;
    b: string;
    a?: string;
}

export interface RGBObject {
    r: number;
    g: number;
    b: number;
    a?: number;
}

export interface RGBObjectGeneric {
    r: NumberOrString;
    g: NumberOrString;
    b: NumberOrString;
    a?: NumberOrString;
}

export interface HSLObject {
    h: number;
    s: number;
    l: number;
    a?: number;
}

export interface HSLObjectGeneric {
    h: number;
    s: NumberOrString;
    l: NumberOrString;
    a?: number;
}

export interface CMYKObject {
    c: number;
    m: number;
    y: number;
    k: number;
    a?: number;
}

export interface CMYKObjectGeneric {
    c: NumberOrString;
    m: NumberOrString;
    y: NumberOrString;
    k: NumberOrString;
    a?: number;
}

export interface RYBObject {
    r: number;
    y: number;
    b: number;
    a?: number;
}

export type RGYBObject = RGBObject | RYBObject;

export type Color = RGBObjectGeneric | HSLObjectGeneric | CMYKObjectGeneric;
export type ColorWithoutCMYK = RGBObjectGeneric | HSLObjectGeneric;

export type ColorInput = string | Color;
export type ColorInputWithoutCMYK = string | ColorWithoutCMYK;
export type HEXOutput = string | HEXObject;
export type RGBOutput = string | RGBObject;
export type HSLOutput = string | HSLObject;
export type CMYKOutput = string | CMYKObject;
export type ColorOutput = HEXOutput | RGBOutput | HSLOutput;

export interface ObjectProps<T> {
    [key: string]: T;
}

export interface Options {
    decimals: number;
    legacyCSS: boolean;
    spacesAfterCommas: boolean;
}

export type InputOptions = Partial<Options>;