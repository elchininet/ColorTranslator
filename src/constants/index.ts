export * from './numbers';
export * from './enums';
export * from './regexps';
export * from './errors';
