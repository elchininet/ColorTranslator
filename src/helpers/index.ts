import {
    Options,
    InputOptions,
    NumberOrString,
    ColorInput
} from '@types';
import {
    PCENT,
    HEX,
    MAX_DECIMALS,
    DEFAULT_OPTIONS,
    TypeOf
} from '#constants';

//---Has property
export const hasProp = <T>(obj: T, prop: string): boolean => Object.prototype.hasOwnProperty.call(obj, prop);

//---Get percent number
export const percentNumber = (percent: NumberOrString): number => {
    return +`${percent}`.replace(PCENT, '$1');
};

//---Get a percentage
export const percent = (percent: NumberOrString): number => PCENT.test(`${percent}`)
    ? percentNumber(percent)
    : Math.min(+percent, 100);

//---Convert to decimal
export const getDEC = (hex: string): number => {
    if (hex.length === 1) { hex += hex; }
    return parseInt(hex, 16);
};

//---Convert to hexadecimal
export const getHEX = (number: NumberOrString): string => {
    const hex = round(number, 0).toString(16).toUpperCase();
    if (hex.length === 1) { return `0x0${hex}`; }
    return `0x${hex}`;
};

//---Convert to final hexadecimal
export const toHEX = (h: NumberOrString): string => {
    let hex = round(h, 0).toString(16).toUpperCase();
    if (hex.length === 1) {
        hex = `0${hex}`;
    }
    return hex;
};

//---Calculate a decimal 255 from an RGB color
export const getBase255Number = (color: string, alpha = false): number => {
    if (!alpha && PCENT.test(color)) {
        return Math.min(255 * percentNumber(color) / 100, 255);
    }
    if (HEX.test(color)) {
        if (color.length === 3) {
            color += color.slice(-1);
        }
        return alpha
            ? round(color) / 255
            : round(color);
    }
    return Math.min(+color, alpha ? 1 : 255);
};

//---Calculate a decimal 0-1 value from CMYK value
export const getCMYKNumber = (color: string): number => Math.min(PCENT.test(color) ? percentNumber(color) / 100 : +color, 1);

//---Return an ordered string from an array values
export const getOrderedArrayString = (keys: string[]): string => [...keys].sort().join('').toUpperCase();

//---Round value
export const round = (value: NumberOrString, decimals = MAX_DECIMALS): number => {
    const exp = Math.pow(10, decimals);
    return Math.round(+value * exp) / exp;
};

//---Minimum and maximum
export const minmax = (n: number, min: number, max: number): number => Math.max(min, Math.min(n, max));

//--Radian to grades
export const grades = (radian: number): number => radian * 180 / Math.PI;

export const parseOptions = (options: Partial<Options>): Options => ({
    ...DEFAULT_OPTIONS,
    ...options
});

interface MatchOptions {
    legacyCSS: number;
    spacesAfterCommas: number;
}

export const getOptionsFromColorInput = (options: InputOptions, ...colors: ColorInput[]): Options => {
    const cssColors = colors.filter((color: ColorInput): boolean => typeof color === 'string') as string[];
    const matchOptions: MatchOptions = {
        legacyCSS: 0,
        spacesAfterCommas: 0
    };
    cssColors.forEach((color: string): void => {
        if (color.includes(',')){
            matchOptions.legacyCSS ++;
            const commasWithNextCharacter = color.match(/,( +|\d+)/g);
            if (
                new Set(commasWithNextCharacter).size === 1 &&
                / +/.test(commasWithNextCharacter[0].slice(1))
            ) {
                matchOptions.spacesAfterCommas ++;
            }
        }
    });
    return {
        decimals: typeof options.decimals === TypeOf.NUMBER
            ? options.decimals
            : DEFAULT_OPTIONS.decimals,
        legacyCSS: typeof options.legacyCSS === TypeOf.BOOLEAN
            ? options.legacyCSS
            : Boolean(
                cssColors.length &&
                matchOptions.legacyCSS === cssColors.length
            ) || DEFAULT_OPTIONS.legacyCSS,
        spacesAfterCommas: typeof options.spacesAfterCommas === TypeOf.BOOLEAN
            ? options.spacesAfterCommas
            : Boolean(
                cssColors.length &&
                matchOptions.spacesAfterCommas === cssColors.length
            ) || DEFAULT_OPTIONS.spacesAfterCommas
    };
};