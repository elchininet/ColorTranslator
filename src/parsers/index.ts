export { HEXStringParser } from './_HEXStringParser';
export { RGBStringParser } from './_RGBStringParser';
export { HSLStringParser } from './_HSLStringParser';
export { HWBStringParser } from './_HWBStringParser';
export { CIELabStringParser } from './_CIELabStringParser';
export { CMYKStringParser } from './_CMYKStringParser';