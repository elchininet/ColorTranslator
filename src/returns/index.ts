import {
    Color,
    ColorInput,
    ColorInputWithoutCMYK,
    ColorOutput,
    InputOptions,
    Options,
    RGBObject
} from '@types';
import {
    ColorModel,
    DEFAULT_BLEND_STEPS,
    Harmony,
    HarmonyString,
    Mix,
    MixString
} from '#constants';
import * as utils from '#color/utils';
import { getOptionsFromColorInput, getRGBObject } from '#color/extractors';

export const getColorReturn = <T>(
    color: ColorInput,
    model: ColorModel,
    decimals: number,
    translateFunction: (color: Color, decimals: number) => T
): T => {
    const rgbObject = getRGBObject(color, model);
    return translateFunction(rgbObject, decimals);
};

export const getBlendReturn = <T>(
    from: ColorInput,
    to: ColorInput,
    steps: number,
    decimals: number,
    translateFunction: (color: Color, decimals: number) => T
): T[] => {
    if (steps < 1) steps = DEFAULT_BLEND_STEPS;
    const fromRGBObject = getRGBObject(from);
    const toRGBObject = getRGBObject(to);
    const blendArray = utils.blend(fromRGBObject, toRGBObject, steps);
    return blendArray.map((color: RGBObject): T => {
        return translateFunction(color, decimals);
    });
};

interface GetBlendReturnObjectWithParameters<T>{
    from: ColorInput;
    to: ColorInput;
    thirdParameter?: number | InputOptions;
    fourthParameter?: InputOptions;
    translateFunction: (color: Color, decimals: number) => T;
}

interface GetBlendReturnStringWithParameters<T> extends GetBlendReturnObjectWithParameters<T> {
    cssFunction?: (color: T, options: Options) => string;
}

export function getBlendReturnWithParameters<T>(
    params: GetBlendReturnObjectWithParameters<T>
): T[];
export function getBlendReturnWithParameters<T>(
    params: GetBlendReturnStringWithParameters<T>
): string[];
export function getBlendReturnWithParameters<T>(
    params: GetBlendReturnStringWithParameters<T>
): T[] | string[] {
    const {
        from,
        to,
        thirdParameter,
        fourthParameter,
        translateFunction,
        cssFunction
    } = params;
    const result = typeof thirdParameter === 'number'
        ? getBlendReturn<T>(
            from,
            to,
            thirdParameter,
            fourthParameter?.decimals,
            translateFunction
        )
        : getBlendReturn<T>(
            from,
            to,
            DEFAULT_BLEND_STEPS,
            thirdParameter?.decimals,
            translateFunction
        );
    if (cssFunction) {
        return result.map((color: T): string => cssFunction(
            color,
            getOptionsFromColorInput(
                (
                    typeof thirdParameter === 'number'
                        ? fourthParameter
                        : thirdParameter
                ) ?? {},
                from,
                to
            )
        ));
    }
    return result;

}

interface GetMixReturn {
    colors: ColorInput[];
    secondParameter?: MixString | InputOptions;
    thirdParameter?: InputOptions;
    colorMixerFunction: (color: ColorInput[], mix: MixString, isCSS: boolean, options: Options) => string;
}

interface GetMixObjectReturn extends GetMixReturn {
    css: false;
}

interface GetMixCssReturn extends GetMixReturn {
    css: true;
}

export function getMixReturn<T>(params: GetMixObjectReturn): T;
export function getMixReturn(params: GetMixCssReturn): string;
export function getMixReturn<T>(
    params: GetMixObjectReturn | GetMixCssReturn
): T | string {
    const {
        colors,
        secondParameter,
        thirdParameter,
        colorMixerFunction,
        css
    } = params;
    const options = getOptionsFromColorInput(
        (
            typeof secondParameter === 'string'
                ? thirdParameter
                : secondParameter
        ) ?? {},
        ...colors
    );
    if (typeof secondParameter === 'string') {
        return colorMixerFunction(
            colors,
            secondParameter,
            css,
            options
        );
    }
    return colorMixerFunction(
        colors,
        Mix.ADDITIVE,
        css,
        options
    );
}

export const getHarmonyReturn = (
    harmony: HarmonyString,
    color: ColorInputWithoutCMYK,
    mode: MixString,
    options: Options
): ColorOutput[] => {
    return ({
        [Harmony.ANALOGOUS]:           utils.colorHarmony.buildHarmony(color, utils.analogous, mode, options),
        [Harmony.COMPLEMENTARY]:       utils.colorHarmony.buildHarmony(color, utils.complementary, mode, options),
        [Harmony.SPLIT_COMPLEMENTARY]: utils.colorHarmony.buildHarmony(color, utils.splitComplementary, mode, options),
        [Harmony.TRIADIC]:             utils.colorHarmony.buildHarmony(color, utils.triadic, mode, options),
        [Harmony.TETRADIC]:            utils.colorHarmony.buildHarmony(color, utils.tetradic, mode, options),
        [Harmony.SQUARE]:              utils.colorHarmony.buildHarmony(color, utils.square, mode, options)
    })[harmony];
};