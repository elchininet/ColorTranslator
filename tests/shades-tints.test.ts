import { ColorTranslator } from '../src';
import { HEXObject, RGBObject, HSLObjectGeneric } from '../src/@types';

describe('ColorTranslator shades and tints tests', (): void => {

    const hexColorFunctions = [
        ColorTranslator.toHEX,
        ColorTranslator.toHEXA
    ];

    const colorFunctions = [
        ColorTranslator.toRGB,
        ColorTranslator.toRGBA,
        ColorTranslator.toHSL,
        ColorTranslator.toHSLA
    ];

    const base = '#FF0000';
    const shades_results = ['#D50000', '#AA0000', '#800000', '#550000', '#2B0000'];
    const tints_results = ['#FF2B2B', '#FF5555', '#FF8080', '#FFAAAA', '#FFD5D5'];

    hexColorFunctions.forEach((fn): void => {
        const inputCSS = fn(base);
        const inputObject = fn(base, false) as (HEXObject & RGBObject & HSLObjectGeneric);
        const alpha = '80';
        const inputCSSWithAlpha = fn(base + alpha);
        const inputObjectWithAlpha = fn(base + alpha, false) as (HEXObject & RGBObject & HSLObjectGeneric);
        it(`Shades tests from ${JSON.stringify(inputCSS)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color));
            const shades = ColorTranslator.getShades(inputCSS, 5);
            expect(shades).toMatchObject(output);
        });
        it(`Shades tests from ${JSON.stringify(inputObject)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color, false));
            const shades = ColorTranslator.getShades(inputObject, 5) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });
        it(`Shades tests with alpha from ${JSON.stringify(inputCSSWithAlpha)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color + alpha));
            const shades = ColorTranslator.getShades(inputCSSWithAlpha, 5);
            expect(shades).toMatchObject(output);
        });
        it(`Shades tests with alpha from ${JSON.stringify(inputObjectWithAlpha)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color + alpha, false));
            const shades = ColorTranslator.getShades(inputObjectWithAlpha, 5) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });

        it(`Tints tests from ${JSON.stringify(inputCSS)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color));
            const shades = ColorTranslator.getTints(inputCSS, 5);
            expect(shades).toMatchObject(output);
        });
        it(`Tints tests from ${JSON.stringify(inputObject)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color, false));
            const shades = ColorTranslator.getTints(inputObject, 5) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });
        it(`Tints tests with alpha from ${JSON.stringify(inputCSSWithAlpha)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color + alpha));
            const shades = ColorTranslator.getTints(inputCSSWithAlpha, 5);
            expect(shades).toMatchObject(output);
        });
        it(`Tints tests with alpha from ${JSON.stringify(inputObjectWithAlpha)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color + alpha, false));
            const shades = ColorTranslator.getTints(inputObjectWithAlpha, 5) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });
    });

    colorFunctions.forEach((fn): void => {
        const inputCSS = fn(base, true, 0);
        const inputObject = fn(base, false, 0) as (HEXObject & RGBObject & HSLObjectGeneric);
        const alpha = '80';
        const inputCSSWithAlpha = fn(base + alpha, true, 0);
        const inputObjectWithAlpha = fn(base + alpha, false, 0) as (HEXObject & RGBObject & HSLObjectGeneric);
        it(`Shades tests from ${JSON.stringify(inputCSS)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color, true, 0));
            const shades = ColorTranslator.getShades(inputCSS, 5, 0);
            expect(shades).toMatchObject(output);
        });
        it(`Shades tests from ${JSON.stringify(inputObject)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color, false, 0));
            const shades = ColorTranslator.getShades(inputObject, 5, 0) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });
        it(`Shades tests with alpha from ${JSON.stringify(inputCSSWithAlpha)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color + alpha, true, 0));
            const shades = ColorTranslator.getShades(inputCSSWithAlpha, 5, 0);
            expect(shades).toMatchObject(output);
        });
        it(`Shades tests with alpha from ${JSON.stringify(inputObjectWithAlpha)}`, (): void => {
            const output = shades_results.map((color: string) => fn(color + alpha, false, 0));
            const shades = ColorTranslator.getShades(inputObjectWithAlpha, 5, 0) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });

        it(`Tints tests from ${JSON.stringify(inputCSS)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color, true, 0));
            const shades = ColorTranslator.getTints(inputCSS, 5, 0);
            expect(shades).toMatchObject(output);
        });
        it(`Tints tests from ${JSON.stringify(inputObject)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color, false, 0));
            const shades = ColorTranslator.getTints(inputObject, 5, 0) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });
        it(`Tints tests with alpha from ${JSON.stringify(inputCSSWithAlpha)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color + alpha, true, 0));
            const shades = ColorTranslator.getTints(inputCSSWithAlpha, 5, 0);
            expect(shades).toMatchObject(output);
        });
        it(`Tints tests with alpha from ${JSON.stringify(inputObjectWithAlpha)}`, (): void => {
            const output = tints_results.map((color: string) => fn(color + alpha, false, 0));
            const shades = ColorTranslator.getTints(inputObjectWithAlpha, 5, 0) as (HEXObject | RGBObject | HSLObjectGeneric)[];
            expect(shades).toMatchObject(output);
        });
        it(`Shades tests from ${JSON.stringify(inputCSS)} with decimals`, (): void => {
            const shades = ColorTranslator.getShades(inputCSS, 5, 0);
            expect(shades).toMatchSnapshot();
        });
    });

});