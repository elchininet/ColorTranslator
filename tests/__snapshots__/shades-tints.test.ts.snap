// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorTranslator shades and tints tests Shades tests from "hsl(0,100%,50%)" with decimals 1`] = `
[
  "hsl(0,100%,42%)",
  "hsl(0,100%,33%)",
  "hsl(0,100%,25%)",
  "hsl(0,100%,17%)",
  "hsl(0,100%,8%)",
]
`;

exports[`ColorTranslator shades and tints tests Shades tests from "hsla(0,100%,50%,1)" with decimals 1`] = `
[
  "hsla(0,100%,42%,1)",
  "hsla(0,100%,33%,1)",
  "hsla(0,100%,25%,1)",
  "hsla(0,100%,17%,1)",
  "hsla(0,100%,8%,1)",
]
`;

exports[`ColorTranslator shades and tints tests Shades tests from "rgb(255,0,0)" with decimals 1`] = `
[
  "rgb(213,0,0)",
  "rgb(170,0,0)",
  "rgb(128,0,0)",
  "rgb(85,0,0)",
  "rgb(43,0,0)",
]
`;

exports[`ColorTranslator shades and tints tests Shades tests from "rgba(255,0,0,1)" with decimals 1`] = `
[
  "rgba(213,0,0,1)",
  "rgba(170,0,0,1)",
  "rgba(128,0,0,1)",
  "rgba(85,0,0,1)",
  "rgba(43,0,0,1)",
]
`;
